overlord/state/change_test.go

```diff
@@ -1119,3 +1119,347 @@ func (ts *taskRunnerSuite) TestCheckTaskDependencies(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("change", "...")
+
+	t1 := st.NewTask("task1", "...")
+	t2 := st.NewTask("task2", "...")
+	t3 := st.NewTask("task3", "...")
+	t4 := st.NewTask("wait-task", "...")
+
+	// Setup circular dependency between t1,t2 and t3, they should
+	// still act normally.
+	t2.WaitFor(t1)
+	t3.WaitFor(t2)
+	t1.WaitFor(t3)
+
+	chg.AddTask(t1)
+	chg.AddTask(t2)
+	chg.AddTask(t3)
+	chg.AddTask(t4)
+
+	// To trigger the cyclic dependency check, we must trigger the isWaiting logic
+	// and we do this by putting t4 into WaitStatus.
+	t4.SetToWait(state.DoneStatus)
+
+	// task1 (do) => task2 (do) => task3 (do) no reboot
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) => task2 (do) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoingStatus)
+	c.Check(chg.Status(), Equals, state.DoingStatus)
+
+	// task1 (wait) => task2 (do) => task3 (do) means need a reboot
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+
+	// task1 (done) => task2 (wait) => task3 (do) means need a reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+}
+
+func (cs *changeSuite) TestIsWaitingMultipleDependencies(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("change", "...")
+
+	t1 := st.NewTask("task1", "...")
+	t2 := st.NewTask("task2", "...")
+	t3 := st.NewTask("task3", "...")
+	t4 := st.NewTask("wait-task", "...")
+	t3.WaitFor(t1)
+	t3.WaitFor(t2)
+
+	chg.AddTask(t1)
+	chg.AddTask(t2)
+	chg.AddTask(t3)
+	chg.AddTask(t4)
+
+	// Put t4 into wait-status to trigger the isWaiting logic each time
+	// for the change.
+	t4.SetToWait(state.DoneStatus)
+
+	// task1 (do) + task2 (do) => task3 (do) no reboot
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (done) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (do) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// For the next two cases we are testing that a task with dependencies
+	// which have completed, but in a non-successful way is handled correctly.
+	// The reason we expect a DoStatus from the following:
+	// - Task 1 or 2 is in error
+	// - Task 1 or 2 is wait
+	// - Task 3 depends on 1+2, and is in Do.
+	// - Task 4, is independent, and in Wait.
+	// It may seem logically that this change would be in error or wait mode, but because of
+	// our status ordering, Do takes precedence above Wait and Error, so it's important
+	// that our WaitStatus logic (which actually precedes the ordering) is respecting this
+	// by determining the next two cases to be identical, and not opinionated otherwise.
+	// task1 (error) + task2 (wait) => task3 (do) no reboot
```

Comment:
@alfonsosanchezbeato @MiguelPires I'm not sure I agree with this comment at the end of the day the change will be stuck until the task2 status changes to something different from wait, so I'm not sure why we wouldn't want to ignore error (as a ready status) and consider the change wait

Reply:
Yes, I actually agree with that, it should be consider a WaitStatus. I think that the status ordering in `statusOrder` cannot be applied blindly in all situations anymore, and that we should start to consider the full task graph to set a state for the change instead. Probably a diagram with the state transitions would help to understand this better.

I am also thinking that some situations are probably not possible, aren't for instance all tasks in Do transitioned to Hold in the moment one task is set to Error? If tat is the case the case described here would not really happen in reality.